### PR TITLE
GunzipContentProvider implementation without Content.Processor

### PR DESCRIPTION
**Pull request by lorban**
**Jetty version**
12

The simplest possible `Content.Provider` implementation I could think of that gunzips content sourced from another `Content.Provider`.

See #7407.

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/GunzipContentProvider.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            inflatedBuffer = null;
+            currentDeflatedContent.release();
+            currentDeflatedContent = null;
+        }
+    }
+
+    @Override
+    public void demandContent(Runnable onContentAvailable)
+    {
+        deflatedProvider.demandContent(onContentAvailable);
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/GunzipContentProvider.java:85]: I think you're covering only the case where the user of this class uses the style:
```java
onContentAvailable() {
  while (true) {
    Content c = readContent();
    if (c == null)
      demandContent(this::onContentAvailable);
  } else {
    consumeContent(c);
}
```

But there is also the style:

```java
onContentAvailable() {
  Content c = readContent();
  consumeContent(c);
  demandContent(this::onContentAvailable);
}
```
I don't think the second style will work if you always forward the demand upstream, right?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/GunzipContentProvider.java (hunk 1 above)
Comment (lorban) [reply jetty-server/src/main/java/org/eclipse/jetty/server/GunzipContentProvider.java:85]: You're right, but that's a different problem that can be solved independently. Let me add an extra test for this use-case as well as the required implementation.

**Comment by lorban** [comment]
Something important I noticed about the `Provider` interface's contract: the `void demandContent(Runnable onContentAvailable)` method must be thread-safe while the `Content readContent()` one must not.

This must be carefully documented (or discussed in case we disagree on this).

File: jetty-server/src/main/java/org/eclipse/jetty/server/GunzipContentProvider.java (hunk 1 above)
Comment (lorban) [reply jetty-server/src/main/java/org/eclipse/jetty/server/GunzipContentProvider.java:85]: @sbordet I've tackled this problem with the simplest possible implementation I could think of (again), without modifying anything related to `readContent()`.

The only thing worth mentioning I relied upon is that `Provider.demandContent()` is thread-safe while `Provider.readContent()` is not.

**Comment by gregw** [comment]
There are 3 "events": a call to `demandContent(Runnable)`; a call to `readContent()`; and the arrival of data available notification that can happen any time during/after `demandContent()`.
It is unlikely, but not impossible for the app to call `demandContent` and/or `readContent` concurrently, so the main race is in the arrival of available notification in the implementation of `demandContent`.  All other races could be said to be bad applications, but we should strive to behaive nicely none the less.

**Comment by gregw** [comment]
Oh and the secondary race of available notification or `demandContent` call   racing with the return of a call to the Runnable passed to the previous `demandContent` call.

**Pull request closed without merging**